src/main/scala/com/evolutiongaming/cluster/sharding/ShardingStrategy.scala

```diff
@@ -70,6 +70,56 @@ object ShardingStrategy {
   }
 
 
+  /**
+    * Per-shard allocation and rebalance based on matching shard's "role" with node roles
+    */
+  class FilterByRole(
+    shardRole: Shard => Option[String],
+    toAddress: Region => Address,
+    clusterMembersWithRoles: => Map[Address, Set[String]],
+    strategy: ShardingStrategy) extends ShardingStrategy {
+
+    def allocate(requester: Region, shard: Shard, current: Allocation): Option[Region] = {
+
+      shardRole(shard) match {
+        case Some(role) =>
+
+          val included = for {
+            (memberAddress, roles) <- clusterMembersWithRoles if roles contains role
+            region <- current.filter { case (region, _) => toAddress(region) == memberAddress }
+          } yield region
+
+          if (included.isEmpty) None
+          else {
+            val region = strategy.allocate(requester, shard, included)
+            region flatMap { region =>
+              if (included contains region) Some(region) else included.keys.headOption
```

Comment:
Interesting case, just curious about your motivation to proceed this way?

Reply:
Yes, it's possible to first check if `requester` is in `included` and use it if yes, and then fallback to  `included.keys.headOption` if not.